### PR TITLE
Dynamic cost model to mitigate high E2E perf variance 

### PR DESCRIPTION
**Pull request by RandySheriffH**
Addressing [issue](https://microsoft-my.sharepoint.com/personal/jianm_microsoft_com/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fjianm%5Fmicrosoft%5Fcom%2FDocuments%2FMicrosoft%20Teams%20Chat%20Files%2FOnnxRuntimePerfEvaluation%2Ehtml&parent=%2Fpersonal%2Fjianm%5Fmicrosoft%5Fcom%2FDocuments%2FMicrosoft%20Teams%20Chat%20Files)

Research showed that the constant cost model now utilized in thread pool split tasks with big granularity, sometimes this helps to infer the model faster,  while sometimes it is causing high E2E latency since occasionally there is a thread, after acquiring a large shard of task, get preempted from execution.

To battle the problem, we devised an approach to compute the block size on the fly - the block size starts from 
```
total_number_of_iterations /  (num_of_threads * dynamic_block_base)
```
And it will diminish according to the changing residual. This way, when the task is approaching completion, the preempted thread (if there is any) will take a shard that is way smaller, allowing sibling threads to keep contributing for parallelism.
Tests on D32s (with 13 threads) suggest variance reduction as below: 

| Case | Mean(ms) | Min(ms) | Max(ms) | Std | Variance |
| ----- |----- | ----- |----- |----- | ----- |
| resnet with constant cost model | 8.38 | 7.66 | 25.49 | 0.58 | 0.073 |
| resnet with dynamic cost model, base 2  | 8.48 | 7.71 | 20.09 | 0.36 | 0.046 |
| resnet with dynamic cost model, base 4  | 8.46 | 7.67 | 16.75 | 0.26 | 0.032 |
| mobilenet with constant cost model | 15.57 | 13.7 | 27.68 | 0.6 | 0.04 |
| mobilenet with dynamic cost model, base 2  | 15.23 | 13.66 | 22.68 | 0.55 | 0.37 |
| mobilenet with dynamic cost model, base 4  | 15.41 | 13.75 | 25.5 | 0.56 | 0.37 |
| ssdmobile with constant cost model | 38.31 | 40.37 | 80.27 | 1.384 | 0.034 |
| ssdmobile with dynamic cost model, base 2 | 40.58 | 38.19 | 64.52 | 1.127 | 0.028  |
| ssdmobile with dynamic cost model, base 4 | 40.85 | 38.49 | 61.25 | 1.026 | 0.025 |

Also, there are significant perf gains for some production models:

| Model | Qps | Mean(ms) | 90 percentile(ms) | 95 percentile(ms) | 
| ----- |----- | ----- |----- |----- |
| prod_deepwrite with constant cost model | 37.46 | 26.70 | 29.36 | 29.77 |
| prod_deepwrite with dynamic cost model, base 4 | 53.64 | 18.64 | 20.17| 20.66 |
| prod_model12 with constant cost model | 398.39 | 2.51 | 3.49 | 3.93 |
| prod_model12 with dynamic cost model, base 4 | 593.47 | 1.69 | 1.89 | 1.97 |
| prod_agiencoder with constant cost model | 653.90 | 1.53 | 2.62 | 2.68 |
| prod_agiencoder with dynamic cost model, base 4 | 839.50 | 1.19 | 1.24 | 1.25 |

Some more numbers from physicals:

Intel(R) Xeon(R) CPU E5-2620, 6 cores
///////////////////// ssdmobile ///////////////////////
const model : qps=22.57, mean=0.0442, std=0.011629260801434909, var=0.00013523970678779048
dynamic 2: qps=24.59, mean=0.0406, std=0.008277092821418143, var=6.851026557437177e-05
dynamic 4: qps=26.44, mean=0.0378, std=0.007924084170619121, var=6.279110994305653e-05
/////////////////// mobilenet //////////////////////////
const model: qps=32.20, mean=0.0310, std=0.0048688660046941395, var=2.3705856171666273e-05
dynamic 2: qps=33.43, mean=0.0298, std=0.004343168162079017, var=1.8863109684096833e-05
dynamic 4: qps=33.61, mean=0.0297, std=0.0037940134098934084, var=1.439453775445101e-05
////////////////////// resnet50 ////////////////////////
const model: qps=17.84, mean=0.0560, std=0.015767396640308858, var=0.00024861079681282305
dynamic 2: qps=19.32, mean=0.0517, std=0.010883550771730607, var=0.00011845167740083788
dynamic 4: qps=18.06, mean=0.0553, std=0.013859506849727157, var=0.00019208593011763397

Intel(R) Xeon(R) Gold 5220R, 24 cores
////////////// ssdmobile /////////////
const model: qps=51.80, mean=0.0193, std=0.00024220350741531048, var=5.866253900427835e-08
dynamic 2: qps=60.09, mean=0.0166, std=0.000614663327857949, var=3.7781100661340857e-07
dynamic 4: qps=50.67, mean=0.0197, std=0.00020406390398409303, var=4.1642076909229135e-08
///////////// mobilenet //////////////
const model: qps=53.93, mean=0.0185, std=0.00016883857704527312, var=2.850646509867263e-08
dynamic 2: qps=53.70, mean=0.0186, std=0.0001712141160763688, var=2.9314273543812286e-08
dynamic 4: qps=53.63, mean=0.0186, std=0.00013711258923688893, var=1.879986212724383e-08
/////////////// resnet50 /////////////////
const model: qps=87.87, mean=0.0113, std=0.00017743540289509786, var=3.14833222005457e-08
dynamic 2: qps=88.05, mean=0.0113, std=0.00014747379494455025, var=2.1748520195347252e-08
dynamic 4: qps=88.00, mean=0.0113, std=0.00011914453556872791, var=1.4195420355887871e-08

### CODE REVIEW

File: onnxruntime/core/common/threadpool.cc
Comment (yuslepukhin) [inline onnxruntime/core/common/threadpool.cc:408]: >berof [](http://example.com/codeflow?start=65&length=5)

split

Comment (RandySheriffH) [reply onnxruntime/core/common/threadpool.cc]: Done.

File: onnxruntime/core/common/threadpool.cc
Comment (tlh20) [inline onnxruntime/core/common/threadpool.cc:428]: Main comment on the code: can we structure this so that the loop itself is the same, but we use a different form of LoopCounter?  We would have

StaticLoopCounter == What we have today
DynamicLoopCounter == This PR

We could then use a C++ template for the loop itself, parameterized by "T", and instantiated once with T=StaticLoopCounter, and once with T=DynamicLoopCounter.

File: onnxruntime/core/common/threadpool.cc
Comment (tlh20) [inline onnxruntime/core/common/threadpool.cc:437]: One consequence of using fetch_add on a single counter here is that we _lose_ the ability to cause the same thread to tend to be allocated the same work across a series of loops (e.g. thread 0 runs iterations 0..9, thread 1 runs iterations 10..19, etc).  That was important for cache locality in some models -- I think it was one using GRU where we first saw that, and e-mail discussion w Tracy and maybe Du Li?

Could we pushing the batch-size selection _inside_ LoopCounter, so that we can continue using the same sharding work there to provide this locality?

That will also help avoid needing two different kinds of loop -- we would just have a change _inside_ LoopCounter to manage the dynamic batch size selection.

**Review by tlh20 — CHANGES_REQUESTED** [review]
Really like the idea of trying guided self scheduling!  As discussed on Teams chat, there may be other choices to pick up from the paper before settling on the final algorithm.  Main comments:

* In terms of code, hopefully we can encapsulate the choice of batch size into LoopCounter.  That will help us combine this work with the steps taken to maintain affinity between threads and loop iterations (at least for the majority of the loop, in the case where we don't have thread pre-emption / oversubscription / etc).

* Can you sync with Pranav on how much perf testing is required before merging?  Maybe we can schedule a session in the AIFX Tech Council call to discuss (Nick is setting these up again in the new year).  The main results in the PR look good, but it is one specific test case in terms of #threads and machine size... think we need a robust / standard mix of workloads to test each time.  We're getting a more complex mix of h/w, plus features like asymmetric core performance in Alder Lake, so it's important to check whether an improvement in one place is balanced by a regression somewhere else.

File: onnxruntime/core/util/thread_utils.h
Comment (RandySheriffH) [reply onnxruntime/core/util/thread_utils.h]: Done.

File: onnxruntime/core/common/threadpool.cc
Comment (RandySheriffH) [reply onnxruntime/core/common/threadpool.cc]: num_of_blocks = d_of_p * dynamic_block_base;
when dynamic_block_base set to 1, we will have:
num_of_blocks = d_of_p
But if, for some case, we prefer smaller granularity, it is attainable by changing the dynamic_block_base.

File: onnxruntime/core/common/threadpool.cc
Comment (RandySheriffH) [reply onnxruntime/core/common/threadpool.cc]: Refactor a bit on current LoopCounter which accepts a block_size on ClaimIterations.

File: onnxruntime/core/common/threadpool.cc
Comment (RandySheriffH) [reply onnxruntime/core/common/threadpool.cc]: With latest commits I am reusing the LoopCounter for better cache locality.

File: onnxruntime/test/platform/threadpool_test.cc
Comment (RandySheriffH) [reply onnxruntime/test/platform/threadpool_test.cc]: Sure - test cases expanded.

File: onnxruntime/core/common/threadpool.cc
Comment (tlh20) [inline onnxruntime/core/common/threadpool.cc:309]:
```suggestion
                       uint64_t& my_end, 
                       uint64_t block_size) {
```

[ adjusting indentation to match surrounding items ]

File: onnxruntime/core/common/threadpool.cc
Comment (tlh20) [inline onnxruntime/core/common/threadpool.cc:440]:
```suggestion
        // Re-compute the amount of work remaining, and adjust the
        // block size if appropriate.
        // TODO : If we make dynamic sizing the default, then we should
        // investigate techniques to avoid an atomic here.  For instance, we
        // could step down the next block size algorithmically based on the
        // amount of work that we have claimed.
        auto todo = left.fetch_sub(my_iter_end - my_iter_start, std::memory_order_relaxed);
        if (b > 1) b = std::max(1LL, std::llroundl(static_cast<long double>(todo) / num_of_blocks));
```

**Review by tlh20 — APPROVED** [review]
Thanks for the updates, and extensions to the test cases.  I added a couple of comments, but not to block merging.  The main technical point -- if we get good perf results and want to adopt this as the default -- is we should look at removing the additional atomic used in work distribution.  I added a TODO as a comment to remind.

Prior to merging, can you sync with Pranav on any additional testing he would like done (if you guys didn't get a chance to discuss already before the holidays).  The range of results is looking good from my viewpoint, but I'd like to make sure we are consistent across perf-related PRs.

File: onnxruntime/core/common/threadpool.cc
Comment (RandySheriffH) [reply onnxruntime/core/common/threadpool.cc]: Done.

File: onnxruntime/core/util/thread_utils.h
Comment (pranavsharma) [inline onnxruntime/core/util/thread_utils.h:21]:
```suggestion
  //It it is non-negative, thread pool will split a task by a decreasing block size
```

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:106]: Can we also document what's a good value to use?

**Comment by pranavsharma** [comment]
We also need to document this setting in the performance tuning doc with clear explanation of how to go about using (tuning) it.

File: onnxruntime/core/util/thread_utils.h
Comment (RandySheriffH) [reply onnxruntime/core/util/thread_utils.h]: Done.

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h
Comment (RandySheriffH) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:106]: Yes - plan to add a section in "perf-tuning" part of the doc, to explain the usage and possible impact.

**Comment by RandySheriffH** [comment]
> I'll sync up offline regarding the tests. Basically we run a bunch of models in Anubis. Have we run the same set of models with this turned on? Were there models that were adversely affected; hopefully we can explain that?

Yes we did - 
dynamic cost model : https://anubistest.azurewebsites.net/#/jobDetail/b52315b5-3f30-4cb8-9484-c1e5855dac3e
master: https://anubistest.azurewebsites.net/#/jobDetail/5334e451-d18e-41aa-8584-c679b00a1991

There are a few of models that are adversely affected such as prod_deepthink_v2 and prod_keypoints, the explanation is that these models prefer to use SimpleParallelFor with minimized black size (which is 1) for better parallelism; However the dynamic model suggested in the PR utilize way bigger block size.

**Review by pranavsharma — APPROVED** [review]

**Pull request merged**
